### ISSUE TITLE
Finish up `use_heap` feature.

### ISSUE THREAD

**Issue by briansmith** [briansmith/ring#78]
- [x] Replace the non-default `no_heap` feature with a default `use_heap` feature.
- [ ] Convert tests from using `std::string::String` and `std::vec::Vec` to use arrayvec's `ArrayString` and `ArrayVec`, respectively.
- [ ] [On Hold] Add `use_heap` configurations to CI configurations.

**Comment by bluss** [comment]
Note that cargo features are supposed to be additive, and a feature turning something off will not work correctly. If crates A and B both depend on ring, and A and B are compiled in the same project, they get the union of the feature flags .. so if one requests no_heap, both get that, which is why feature flags must be additive. See cargo docs for a bit more info..

**Comment by briansmith** [comment]
Thanks! I did not realize that. I was trying to be like `#![no_std]`.

**Comment by briansmith** [comment]
Reducing the scope of this. Making RSA signature verification work without using the heap is now a separate issue: #102. Similarly, making the ECDSA code work without the heap is now a separate issue: #103.

**Comment by briansmith** [comment]
An update on the plan for this: I'm planning to removing the `use_heap` feature entirely from the default feature set after fixing #102 and #103. After those two features are fixed, RSA signing will be the only feature that uses the heap. Accordingly, I'm planning to add a new `rsa_key_pair` feature that will NOT be a default feature. Also, because RSAKeyPair will be the only use of crypto/bn (`BIGNUM`), we'll avoid even building crypto/bn/* unless the `rsa_key_pair` feature is enabled.

**Comment by briansmith** [comment]
We now do this:

```rust
#[cfg(feature = "use_heap")]
impl std::error::Error for Unspecified {
```

Instead of keying that on `use_heap` we should key it on `std` or `use_std` depending on whatever convention is more common in Rust libraries.

Once #102 is done, we can `#[cfg(not(all(feature = "rsa_signing", feature = "use_heap")))]` with `#[cfg(feature = "rsa_signing")]` and make rsa_signing depend on `use_heap.
